### PR TITLE
Generic HTTP client - feature 23

### PR DESCRIPTION
**Pull request by aydevelop**

### CODE REVIEW

File: frontend/src/environments/environment.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@
 // The list of file replacements can be found in `angular.json`.
 
 export const environment = {
+  apiUrl: 'https://localhost:5001',
```
Comment (Jyuart) [inline frontend/src/environments/environment.ts]:
```suggestion
  apiUrl: 'https://localhost:5001/api',
```

I think we'll have it

File: frontend/src/app/services/resource.service.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  providedIn: 'root',
+})
+export abstract class ResourceService<T> {
+  private readonly APIUrl = environment.apiUrl + this.getResourceUrl();
+
+  constructor(protected httpClient: HttpClient) {}
+
+  abstract getResourceUrl(): string;
+
+  toServerModel(entity: T): string {
```
Comment (Jyuart) [inline frontend/src/app/services/resource.service.ts]:
```suggestion
  request(entity: T): string {
```

File: frontend/src/app/services/resource.service.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    return JSON.stringify(entity);
+  }
+
+  fromServerModel(json: IResponse): T[] {
```
Comment (Jyuart) [inline frontend/src/app/services/resource.service.ts]:
```suggestion
  response(json: IResponse): T[] {
```

Comment (Jyuart) [inline frontend/src/app/services/resource.service.ts]:
```suggestion
  fromServerModel(json: IResponse): T {
```

File: frontend/src/app/services/resource.service.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    let item = JSON.parse(json.data);
+    var data: T[] = Array.isArray(item) ? item : [item];
+    return data;
+  }
+
+  getList(index: number, page: number): Observable<T[]> {
```
Comment (Jyuart) [inline frontend/src/app/services/resource.service.ts]: Not sure we'll need paging everywhere. Please, create overload without those parameters both in function and request

File: frontend/src/app/services/resource.service.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    let params = new HttpParams()
+      .set('limit', index.toString())
+      .set('offset', page.toString());
+
+    return this.httpClient
+      .get<IResponse[]>(`/${this.APIUrl}?${params.toString()}`)
```
Comment (Jyuart) [inline frontend/src/app/services/resource.service.ts]: Only the data in the IReponse may be an array, not the whole response object

File: frontend/src/app/services/resource.service.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      .pipe(
+        switchMap((list) => list.map((item) => this.fromServerModel(item))),
+        catchError(this.handleError)
+      );
+  }
+
+  get(id: string | number): Observable<T> {
```
Comment (Jyuart) [inline frontend/src/app/services/resource.service.ts:37]: I don't think we need two different methods for getting a single resource or a list. Make it just one get

File: frontend/src/app/services/resource.service.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    return this.httpClient.get<IResponse>(`/${this.APIUrl}/${id}`).pipe(
+      switchMap((item) => this.fromServerModel(item)),
+      take(1),
+      catchError(this.handleError)
+    );
+  }
+
+  add(resource: T): Observable<IResponse> {
+    return this.httpClient
+      .post<IResponse>(`/${this.APIUrl}`, this.toServerModel(resource))
```
Comment (Jyuart) [inline frontend/src/app/services/resource.service.ts]:
```suggestion
      .post<T>(`/${this.APIUrl}`, this.toServerModel(resource))
```

You'll send a resource, not a response object.
The same for put and delete

**Review by Jyuart — CHANGES_REQUESTED** [review]

**Comment by aydevelop** [comment]
Top method removed for code cleanliness
Look, generic for response type
Removed pagination, created separate method for it - reload doesn't work like in C#

File: frontend/src/app/services/resource.service.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  get(id: string | number): Observable<T> {
+    return this.httpClient.get<IResponse>(`/${this.APIUrl}/${id}`).pipe(
+      map((item) => JSON.parse(item.data)),
+      catchError(this.handleError)
+    );
+  }
+
+  add(resource: T): Observable<IResponse> {
+    return this.httpClient
+      .post<IResponse>(`/${this.APIUrl}`, JSON.stringify(resource))
```
Comment (Jyuart) [inline frontend/src/app/services/resource.service.ts]: You're posting T (the resource). No need to use the IResponse interface here. It's only for responses coming from the backend

Comment (aydevelop) [reply frontend/src/app/services/resource.service.ts]: I edited the code, but the studio is cursing, any ideas

?

Comment (Jyuart) [reply frontend/src/app/services/resource.service.ts]: Sure, you should specify that you're posting a resource of a specific type: `.post<T>`. In that way you can even remove this part: `Observable<T>`. TypeScript is smart enough to know what to expect.

**Review by Jyuart — APPROVED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Pull request merged**